pkg/runtime/unstructured/converter.go

```diff
@@ -0,0 +1,234 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package unstructured
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/json"
+)
+
+// Converter knows how to convert betweek runtime.Object and
+// Unstructured in both ways.
+type Converter struct {
+}
+
+func NewConverter() *Converter {
+	return &Converter{}
+}
+
+func (c *Converter) FromUnstructured(u map[string]interface{}, obj runtime.Object) error {
+	return c.fromUnstructured(reflect.ValueOf(u), reflect.ValueOf(obj).Elem())
+}
+
+func (c *Converter) fromUnstructured(sv, dv reflect.Value) error {
+	sv = unwrapInterface(sv)
+	st, dt := sv.Type(), dv.Type()
+
+	switch dt.Kind() {
+	case reflect.Map, reflect.Slice, reflect.Ptr, reflect.Struct, reflect.Interface:
+		// Those require non-trivial conversion.
+	default:
+		// This should handle all simple types.
+		if st.AssignableTo(dt) {
+			dv.Set(sv)
+			return nil
+		}
+		if st.ConvertibleTo(dt) {
+			dv.Set(sv.Convert(dt))
+			return nil
+		}
+	}
+
+	// If the from value is string, check if there is custom marshaller
+	// for the out type and try to use it.
+	//if st.Kind() == reflect.String && dv.CanAddr() {
+	if dv.CanAddr() {
+		// Check if the object has a custom JSON marshaller/unmarshaller.
+		unmarshal := dv.Addr().MethodByName("UnmarshalJSON")
+		if unmarshal.IsValid() {
+			// UnmarshalJSON takes []byte as an argument. However, it assumes
+			// that this is json encoded, whereas here it isn't. We need to
+			// encode it first.
+			data, err := json.Marshal(sv.Interface())
+			if err != nil {
+				return fmt.Errorf("Error encoding to json")
+			}
+			ret := unmarshal.Call([]reflect.Value{reflect.ValueOf(data)})[0].Interface()
+			if ret != nil {
+				return ret.(error)
+			}
+			return nil
+		}
+	}
+
+	switch dt.Kind() {
+	case reflect.Map:
+		return c.mapFromUnstructured(sv, dv)
+	case reflect.Slice:
+		return c.sliceFromUnstructured(sv, dv)
+	case reflect.Ptr:
+		return c.pointerFromUnstructured(sv, dv)
+	case reflect.Struct:
+		return c.structFromUnstructured(sv, dv)
+	case reflect.Interface:
+		return c.interfaceFromUnstructured(sv, dv)
+	default:
+		return fmt.Errorf("Unrecognized type: %v", dt.Kind())
+	}
+}
+
+func fieldNameFromField(field *reflect.StructField) string {
+	jsonTag := field.Tag.Get("json")
```

Comment:
this is fairly expensive (or was) - I think we should consider caching the field -> tag mapping sooner rather than later once we have a benchmark that tells us how bad this overall.